### PR TITLE
type _dynamo/profiler.py

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137351

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/profiler.py [hunk 1]
```diff
@@ -41,23 +41,25 @@ def __truediv__(self, other):
...
-    def __init__(self, captured, total, unique_graphs) -> None:
+    def __init__(
+        self, captured: ProfileMetrics, total: ProfileMetrics, unique_graphs: int
+    ) -> None:
         self.captured: ProfileMetrics = captured or ProfileMetrics()
         self.total: ProfileMetrics = total or ProfileMetrics()
         self.unique_graphs: int = unique_graphs
 
-    def __iadd__(self, other: "ProfileResult"):
+    def __iadd__(self, other: "ProfileResult") -> Self:
```
Comment (Skylion007) [inline torch/_dynamo/profiler.py:56]:
```suggestion
    def __iadd__(self, other: Self) -> Self:
```

File: torch/_dynamo/profiler.py [hunk 2]
```diff
@@ -15,21 +15,21 @@ class ProfileMetrics:
     fusions: int = 0
     graphs: int = 0
 
-    def __iadd__(self, other: "ProfileMetrics"):
+    def __iadd__(self, other: "ProfileMetrics") -> Self:
```
Comment (Skylion007) [inline torch/_dynamo/profiler.py:18]:
```suggestion
    def __iadd__(self, other: Self) -> Self:
```

**Review by Skylion007 — APPROVED** [review]

File: torch/_dynamo/profiler.py [hunk 3]
```diff
@@ -15,21 +15,21 @@ class ProfileMetrics:
...
         self.microseconds += other.microseconds
         self.operators += other.operators
         self.fusions += other.fusions
         return self
 
-    def __add__(self, other: "ProfileMetrics"):
+    def __add__(self, other: "ProfileMetrics") -> "ProfileMetrics":
```
Comment (Skylion007) [inline torch/_dynamo/profiler.py:24]: Not sure about where else in this file Self is valid, but feel free to give it a try. :)

Comment (bobrenjc93) [reply torch/_dynamo/profiler.py:24]: Alas this doesn't work

```
  Error (MYPY) [return-value]
    Incompatible return value type (got "ProfileMetrics", expected "Self")

         32  |    def __truediv__(self, other: Any) -> Self:
         33  |        if isinstance(other, int):
         34  |            other = ProfileMetrics(other, other, other)
    >>>  35  |        return ProfileMetrics(
         36  |            self.microseconds / max(1, other.microseconds),
         37  |            self.operators / max(1, other.operators),
         38  |            self.fusions / max(1, other.fusions),
```

**Pull request closed without merging**